ddtrace/internal/runtime/runtime_metrics.py

```diff
@@ -54,26 +59,77 @@ class RuntimeWorker(_worker.PeriodicWorkerThread):
     """
 
     FLUSH_INTERVAL = 10
+    _instance = None
 
-    def __init__(self, dogstatsd_url, flush_interval=None):
+    def __init__(self, tracer=None, dogstatsd_url=None, flush_interval=None):
+        # type: (Optional[ddtrace.Tracer], Optional[str], Optional[float]) -> None
         super(RuntimeWorker, self).__init__(
             interval=flush_interval or float(get_env("runtime_metrics", "interval", default=self.FLUSH_INTERVAL)),
             name=self.__class__.__name__,
         )
-        self._dogstatsd_client = get_dogstatsd_client(dogstatsd_url)
+        self.tracer = tracer or ddtrace.tracer
+        self._dogstatsd_client = get_dogstatsd_client(dogstatsd_url or agent.get_stats_url())
+
         # Initialize collector
         self._runtime_metrics = RuntimeMetrics()
+        self._services = {}
+
+        # Register span hook
+        self.tracer.on_start_span(self._set_language_on_span)
+
+    def _set_language_on_span(self, span):
+        # add tags to root span to correlate trace with runtime metrics
+        # only applied to spans with types that are internal to applications
+        if span.parent_id is None and self.tracer._is_span_internal(span):
+            span.meta["language"] = "python"
+
+    @staticmethod
+    def disable():
+        # type: () -> None
+        if RuntimeWorker._instance is None:
+            return
+
+        RuntimeWorker._instance.stop()
+        RuntimeWorker._instance.join()
+        RuntimeWorker._instance = None
+
+    @staticmethod
+    def enable(tracer=None, dogstatsd_url=None, flush_interval=None):
+        # type: (Optional[ddtrace.Tracer], Optional[str], Optional[float]) -> None
+        if RuntimeWorker._instance is not None:
+            return
+
+        runtime_worker = RuntimeWorker(tracer, dogstatsd_url, flush_interval)
+        runtime_worker.start()
         # force an immediate update constant tags
-        self.update_runtime_tags()
-        # Start worker thread
-        self.start()
+        runtime_worker.update_runtime_tags()
```

Comment:
This is going to hit the same fork and uwsgi issue that we hit with the profiler and the tracer.
`register_at_fork` is not enough to cover those use case.

I'm thinking we need something more abstract to solve this case once and for all. Can we pause this for a few days until I try to hack something around this?
I'd try to factorize out the profiler code so it can be reused by the tracer and this runtime metric instance.

Reply:
What was the issue in those cases?

Reply:
Long story short, the worst case we're trying to handle is uwsgi where it calls `fork()` (no the Python implementation) under the hood, and you have a new Python process with all thread states messed up and your thread stopped 😢